File: ios/Sources/ARYPush/Internal/NotificationCodec.swift
import Foundation

/// Serialises a ``PushNotification`` so it can outlive the process.
///
/// Needed for the pending-open record, which has to survive on disk until host or Dart listeners
/// attach. JSON rather than `NSCoding` because the same shape is read back by the Flutter bridge.
enum NotificationCodec {

    static func encode(_ notification: PushNotification) -> String? {
        var payload: [String: Any] = [
            "id": notification.id,
            "data": notification.data,
            "receivedAt": notification.receivedAt.timeIntervalSince1970,
            "wasForeground": notification.wasForeground
        ]
        if let title = notification.title { payload["title"] = title }
        if let body = notification.body { payload["body"] = body }
        if let imageURL = notification.imageURL { payload["imageUrl"] = imageURL }
        if let thread = notification.threadIdentifier { payload["threadId"] = thread }
        if let category = notification.categoryIdentifier { payload["categoryId"] = category }
        if let actionId = notification.actionId { payload["actionId"] = actionId }

        guard let data = try? JSONSerialization.data(withJSONObject: payload) else {
            PushLogger.warn("Could not encode a notification for persistence")
            return nil
        }
        return String(data: data, encoding: .utf8)
    }

    /// Returns `nil` for anything unreadable rather than throwing into a system callback.
    static func decode(_ raw: String?) -> PushNotification? {
        guard let raw,
              let data = raw.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = json["id"] as? String
        else { return nil }

        let receivedAt = (json["receivedAt"] as? TimeInterval).map(Date.init(timeIntervalSince1970:))
        return PushNotification(
            id: id,
            title: json["title"] as? String,
            body: json["body"] as? String,
            imageURL: json["imageUrl"] as? String,
            data: json["data"] as? [String: String] ?? [:],
            receivedAt: receivedAt ?? Date(),
            threadIdentifier: json["threadId"] as? String,
            categoryIdentifier: json["categoryId"] as? String,
            actionId: json["actionId"] as? String,
            wasForeground: json["wasForeground"] as? Bool ?? false
        )
    }
}

File: android/sdk/src/main/kotlin/com/ary/push/internal/notification/NotificationCodec.kt
package com.ary.push.internal.notification

import com.ary.push.internal.PushJson
import com.ary.push.internal.log.PushLogger
import com.ary.push.model.PushNotification
import org.json.JSONObject

/**
 * Serialises a [PushNotification] so it can cross a process boundary or outlive one.
 *
 * Needed in two places, both of which rule out passing the object itself:
 *
 *  * the click `PendingIntent`, which the system holds on the SDK's behalf and may deliver long
 *    after this process has died;
 *  * the pending-open record, which has to survive on disk until Dart or host listeners attach.
 *
 * JSON rather than `Parcelable` because the same shape is read back by the Flutter bridge.
 */
internal object NotificationCodec {

    fun encode(notification: PushNotification): String = JSONObject().apply {
        put("id", notification.id)
        notification.title?.let { put("title", it) }
        notification.body?.let { put("body", it) }
        notification.imageUrl?.let { put("imageUrl", it) }
        put("data", PushJson.toJsonObject(notification.data))
        put("receivedAt", notification.receivedAt)
        notification.sentAt?.let { put("sentAt", it) }
        notification.channelId?.let { put("channelId", it) }
        notification.collapseKey?.let { put("collapseKey", it) }
        notification.actionId?.let { put("actionId", it) }
        put("wasForeground", notification.wasForeground)
    }.toString()

    /** Returns null for anything unreadable rather than throwing into a system callback. */
    fun decode(raw: String?): PushNotification? {
        val json = PushJson.parseObject(raw) ?: return null
        val id = json.optString("id").takeIf { it.isNotEmpty() } ?: return null
        return try {
            PushNotification(
                id = id,
                title = json.optString("title").takeIf { it.isNotEmpty() },
                body = json.optString("body").takeIf { it.isNotEmpty() },
                imageUrl = json.optString("imageUrl").takeIf { it.isNotEmpty() },
                data = PushJson.flattenToStringMap(json.optJSONObject("data")),
                receivedAt = json.optLong("receivedAt", System.currentTimeMillis()),
                sentAt = json.optLong("sentAt").takeIf { it > 0 },
                channelId = json.optString("channelId").takeIf { it.isNotEmpty() },
                collapseKey = json.optString("collapseKey").takeIf { it.isNotEmpty() },
                actionId = json.optString("actionId").takeIf { it.isNotEmpty() },
                wasForeground = json.optBoolean("wasForeground", false)
            )
        } catch (t: Throwable) {
            PushLogger.w(t) { "Could not decode a stored notification" }
            null
        }
    }
}
